### PR TITLE
Fallback for authentication and move long running tasks to threads

### PR DESCRIPTION
**Pull request by sladomic**

### CODE REVIEW

File: app/src/main/res/layout/activity_authentication.xml [hunk 1]
```diff
@@ -14,4 +14,10 @@
         opencv:show_fps="false"
         opencv:camera_id="any" />
 
+    <ImageView
+        android:layout_width="fill_parent"
+        android:layout_height="fill_parent"
+        android:id="@+id/authentication_animation"
+        android:src="@drawable/authentication_animation"
```
Comment (jo-elimu) [inline app/src/main/res/layout/activity_authentication.xml:21]: Is the authentication_animation file missing from the commit?

File: app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java [hunk 2]
```diff
@@ -58,6 +66,8 @@ protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_authentication);
 
+        authenticationAnimation = (ImageView)findViewById(R.id.authentication_animation);
```
Comment (jo-elimu) [inline app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java:69]: I told the animator to use a 3-4 second duration. Is that okay?

File: app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java [hunk 3]
```diff
@@ -149,23 +189,26 @@ public Mat onCameraFrame(CameraBridgeViewBase.CvCameraViewFrame inputFrame) {
...
-        tensorFlow = trainingHelper.getInitializedTensorFlow();
         ppF = new PreProcessorFactory(getApplicationContext());
         numberOfTries = 0;
         animalOverlay = animalOverlayHelper.createOverlay();
-        if (animalOverlay != null){
+        if (animalOverlay != null) {
             mediaPlayerAnimalSound = MediaPlayer.create(this, getResources().getIdentifier(animalOverlay.getSoundFile(), "raw", getPackageName()));
-            mediaPlayerInstruction.setOnCompletionListener(new MediaPlayer.OnCompletionListener() {
+        }
+        preview.enableView();
```
Comment (jo-elimu) [inline app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java:198]: Enable preview of what?

File: app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java [hunk 4]
```diff
@@ -149,23 +189,26 @@ public Mat onCameraFrame(CameraBridgeViewBase.CvCameraViewFrame inputFrame) {
...
+    }
+
+    private void removeAuthenticationAnimation(){
+        if (authenticationAnimation.getVisibility() == View.VISIBLE){
+            runOnUiThread(new Runnable() {
                 @Override
-                public void onCompletion(MediaPlayer mediaPlayer) {
-                    mediaPlayerAnimalSound.start();
+                public void run() {
+                    authenticationAnimation.setVisibility(View.INVISIBLE);
```
Comment (jo-elimu) [inline app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java:208]: Is white background for the animation okay? Or should it be transparent?

File: app/src/main/java/org/literacyapp/authentication/StudentImageCollectionActivity.java [hunk 5]
```diff
@@ -143,14 +140,14 @@ public Mat onCameraFrame(CameraBridgeViewBase.CvCameraViewFrame inputFrame) {
...
-        long time = new Date().getTime();
+        long currentTime = new Date().getTime();
 
         Rect face = new Rect();
         boolean isFaceInsideFrame = false;
         boolean faceDetected = false;
 
-        if(lastTime + TIMER_DIFF < time){
-            lastTime = time;
+        if(lastTime + TIMER_DIFF < currentTime){
```
Comment (jo-elimu) [inline app/src/main/java/org/literacyapp/authentication/StudentImageCollectionActivity.java]: Add parenthesis: `((lastTime + TIMER_DIFF)`

**Review by jo-elimu — CHANGES_REQUESTED** [review]
Looks good, but I think the @drawable/authentication_animation file is missing?

File: app/src/main/res/layout/activity_authentication.xml (hunk 1 above)
Comment (sladomic) [reply app/src/main/res/layout/activity_authentication.xml:21]: yes, just pushed it

File: app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java (hunk 2 above)
Comment (sladomic) [reply app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java:69]: Yes, in the tests it was always about 3-4 seconds

File: app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java (hunk 3 above)
Comment (sladomic) [reply app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java:198]: This is the OpenCV camera preview which needs to be enabled (this start the whole thread which gives us the camera frames)

File: app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java (hunk 4 above)
Comment (sladomic) [reply app/src/main/java/org/literacyapp/authentication/AuthenticationActivity.java:208]: If we use transparent background then you still see the camera preview

File: app/src/main/java/org/literacyapp/authentication/StudentImageCollectionActivity.java (hunk 5 above)
Comment (sladomic) [reply app/src/main/java/org/literacyapp/authentication/StudentImageCollectionActivity.java]: done

**Pull request merged**